src/java/org/apache/cassandra/db/guardrails/DefaultGuardrail.java

```diff
@@ -0,0 +1,690 @@
+import org.apache.cassandra.tracing.Tracing;
+import org.apache.cassandra.utils.NoSpamLogger;
+
+import static java.lang.String.format;
+import static org.apache.cassandra.utils.Clock.Global.currentTimeMillis;
+
+/**
+ * General class defining a given guardrail that guards against some particular usage/condition.
+ *
+ * <p>Some guardrails only emit warnings when triggered, while others fail the query that triggers them. Some may do one
+ * or the other based on a specific threshold.
+ *
+ * <p>Note that all the defined classes support live updates, which is why each guardrail class constructor takes
+ * suppliers of the condition the guardrail acts on rather than the condition itself. This implies that said suppliers
+ * should be fast and non-blocking to avoid surprises.
+ */
+public abstract class DefaultGuardrail implements Guardrail
+{
+    private static final NoSpamLogger logger = NoSpamLogger.getLogger(LoggerFactory.getLogger(DefaultGuardrail.class),
+                                                                      10, TimeUnit.MINUTES);
+    private static final String REDACTED = "<redacted>";
+
+    /** A name identifying the guardrail. */
+    public final String name;
+
+    /** Whether to throw {@link InvalidRequestException} on {@link this#fail(String)}. */
+    private boolean throwOnFailure = true;
+
+    /** Minimum logging and triggering interval to avoid spamming downstream. */
+    private long minNotifyIntervalInMs = 0;
+
+    /** Time of last warning in milliseconds. */
+    private volatile long lastWarnInMs = 0;
+
+    /** Time of last failure in milliseconds. */
+    private volatile long lastFailInMs = 0;
+
+    DefaultGuardrail(String name)
+    {
+        this.name = name;
+    }
+
+    protected void warn(String fullMessage, String redactedMessage)
+    {
+        if (skipNotifying(true))
+            return;
+
+        logger.warn(fullMessage);
+        // Note that ClientWarn will simply ignore the message if we're not running this as part of a user query
+        // (the internal "state" will be null)
+        ClientWarn.instance.warn(fullMessage);
+        // Similarly, tracing will also ignore the message if we're not running tracing on the current thread.
+        Tracing.trace(fullMessage);
+        for (Guardrails.Listener listener : Guardrails.listeners)
+            listener.onWarningTriggered(name, redactedMessage);
+    }
+
+    protected void warn(String fullMessage)
+    {
+        warn(fullMessage, fullMessage);
+    }
+
+    protected void fail(String fullMessage, String redactedMessage)
+    {
+        if (!skipNotifying(false))
+        {
+            logger.error(fullMessage);
+            // Tracing will ignore the message if we're not running tracing on the current thread.
+            Tracing.trace(fullMessage);
+            for (Guardrails.Listener listener : Guardrails.listeners)
+                listener.onFailureTriggered(name, redactedMessage);
```

Comment:
maybe I missed it, but I don't see listeners; can you explain the motivation behind listeners?

Reply:
Currently there aren't any usages of listeners, that's an extension point for notifying other monitoring systems, so operators can observe triggered guardrails in ways different than just observing logs, and maybe trigger actions on frequent activations of a certain guardrail.

Reply:
I would completely bypass listeners mechanism and I would rather emit a diagnostic event via existing diagnostic framework. That is in my opinion the right way to do it. I am afraid we are plumbing here something brand new.

Reply:
if the core use case is monitoring, then I agree with @smiklosovic that we shouldn't do listeners.

In track_warnings it was very important we had metrics at each guardrail (how often are users trying to read too much data?), and adding this as a listener would be confusing as I have no way to filter... so it is simpler to just add the metrics at the call site (this also allows the metrics to be table level).

> I would rather emit a diagnostic event via existing diagnostic framework

There are other benefits of diagnostic framework that don't exist with the current listener... so for a monitoring use case I agree with emitting a diagnostic event.

Reply:
I'm removing the listeners from the PR. This also gets rid of the redacted messages for external monitoring systems, since by now we don't have any external monitoring. There are a bunch of changes in the `GuardrailsTester` utilities because we were using listeners to verify that clients warnings were triggered.

We can add diagnostic events later, not as part of this prototype. Indeed, according to the CEP document integration with diagnostic events is out of scope so that's something to do once guardrails are merged to trunk.